Only upgrade stats options from admin or cron

nickdaugherty opened the pull request:
Prevents a rush of frontend-writes on high traffic sites during a stats
option upgrade.

All frontend writes should be avoided unless absolutely necessary, as
they can easily cause problems on sites with large traffic

My only concern with this approach is returning 'old' data from the
function between time of JP upgrade and the next cron or admin trigger
of this, which could cause unexpected behavior (hopefully the code
using it is defensive enough to fail gracefully with old options).

Fixes #4104

nickdaugherty committed:
Only upgrade stats options from admin or cron

Prevents a rush of frontend-writes on high traffic sites during a stats
option upgrade.

All frontend writes should be avoided unless absolutely necessary, as
they can easily cause problems on sites with large traffic

My only concern with this approach is returning 'old' data from the
function between time of JP upgrade and the next cron or admin trigger
of this, which could cause unexpected behavior (hopefully the code
using it is defensive enough to fail gracefully with old options).

Fixes #4104

nickdaugherty commented:
cc @kraftbj

jeherve commented on modules/stats.php:
```diff
@@ -181,7 +181,8 @@ function stats_footer() {
 	$options = get_option( 'stats_options' );
 
-	if ( !isset( $options['version'] ) || $options['version'] < STATS_VERSION )
+	if ( ( !isset( $options['version'] ) || $options['version'] < STATS_VERSION ) &&
```
Since you're touching that old code, could you add in a space before `isset` to respect WP coding standards, as well as curly braces?

nickdaugherty committed:
Update old code to WP coding standards

nickdaugherty committed:
Add missing parenthesis

nickdaugherty commented:
@jeherve good call on fixing the coding standards, done.

kraftbj commented on modules/stats.php:
```diff
@@ -181,8 +181,10 @@ function stats_footer() {
 
-	if ( !isset( $options['version'] ) || $options['version'] < STATS_VERSION )
+	if ( ( ! isset( $options['version'] ) || $options['version'] < STATS_VERSION ) &&
+		( is_admin() || ( defined( 'DOING_CRON' ) && DOING_CRON ) ) ) {
```
Throughout `stats.php`, we're assuming the output of will be an array of `stats_get_options` options. If someone installed Jetpack via CLI, I don't think this would ever fire to get the default options, so most of the code would at least throw notices about the undefined indexes until a cron run or someone visited wp-admin.

Could spin out a "stats_default_options" that returns the default array, `if ( false === $options ) { return stats_default_options()}` and use that same function in `stats_upgrade_options` to keep things DRY.

Thoughts?

jeherve replied:
@nickdaugherty Did you have a chance to look at @kraftbj's points above?

nickdaugherty replied:
That's a good approach - I'll work on implementing it.

gravityrail commented:
Bumped to 4.4 to clear path for 4.2 to get goldmastered :)

nickdaugherty committed:
Move default stats options to a reusable function

nickdaugherty committed:
Coding standards

nickdaugherty committed:
Only persist upgraded options in admin or cron

This allows the options to be 'upgraded' like normal, as the code
expects, but not persisted to the db and XMLRPC call unless we're not
on the frontend

This means that options will never be out of date during a JP upgrade,
the upgrade will just be recalculated on every request (it's cheap)
until the option gets written by an admin or cron request.

nickdaugherty commented:
I came up with a slightly different approach - allow the stats upgrade code to run every time `stats_get_options()` is called (it's cheap), but only write to the DB and XMLRPC if it's not a frontend request.

This has the benefit of ensuring the value returned from `stats_get_options()` is always in the current format for this JP version while sidestepping the need to return default or outdated stats option arrays, and prevents frontend db/network writes.

samhotchkiss commented:
Hey @nickdaugherty -- I'm bumping this out of the pipeline, but happy to get it in-- drop us a ping in Slack if you'd like to get it in for 4.5 next month

joshbetz commented:
Could we revisit this for a future release?

jeherve commented:
@joshbetz Could you rebase so tests can run again? Thanks!

oskosk commented:
@joshbetz @nickdaugherty is any of you taking care of this PR? Does it still apply a year and a half after the changes were proposed ?

joshbetz commented:
The underlying issue of #4104 still needs to be fixed.

oskosk commented:
@joshbetz I understand. Are you planning to fix it taking over this PR? It needs a rebase. 

The specifity of the problem reported in #4104 like "Be a high traffic, highly concurrent site" is not super easy to reproduce. "Enable a new version of Jetpack" is also somehow vague given that we are almost 15 versions ahead of that report. It would be useful to have instructions in this same PR on how to be sure that we're addressing a problem the we can reproduce.

joshbetz commented:
The problem is that this front-end options update can cause many database writes, impacting server load and therefore performance. If 1000 requests come in at the same time, there are 1000 `UPDATE` queries. By limiting this update to only logged-in requests, it limits the effects of the race condition.

oskosk commented:
I understand that problem as it's specified in this PR's descripton.

Have you confirmed that the changes in this PR fix that problem and do not have any influence on anything that we currently need working ? I'm only asking because you reopened the PR and restored the branch. If this weren't the case I wouldn't be asking.

That's the main problem of the PR being here for more than a year. Apparently no one witnessing the problem has tested it nor confirmed that the changes here do what it's expected from them. 

Also:
* Unit tests are not passing. Looks like they never passed in the first place at least with the last commit of the PR.
* Files are conflicting with latest master and there are no instructions on how to verify that previous behaviour is not affected. 

On #4104 @kraftbj commented:
> It runs on the front-end since we need to know which roles to display, or not, the tracking code. The possible overcounting of views until the next is_admin run (or next cron? not picky) is probably worth the savings here.

... as an explanation of why it's done this way.

Closing the PR does not mean that the issue is closed. But having it open for more than a year and related to an issue make it look as stalled work that could solve the problem but hasn't.

keoshi commented:
Hey @joshbetz @nickdaugherty — can we get an update on this and some answers to @oskosk questions above, please? Thank you both!

kraftbj committed:
Merge branch 'master' into fix/4104-prevent-stats-option-upgrade-from-frontend

gravityrail commented:
Closing this, hasn't been an issue for VIP for a while.
